backend/ttrs/recommend.py

```diff
@@ -1,34 +1,192 @@
+
+    # Collect information from student
+    info['student_grade'] = student.grade
+    info['student_college'] = student.college
+    info['student_department'] = student.department if student.department else None
+    info['student_major'] = student.major if student.major else None
+    info['not_recommends'] = student.not_recommends
+
+    # print(info)
     recommends = []
-    num_candidates = 50
-    num_recommends = 3
-    for i in range(num_candidates):
-        print('table', i)
-        time_table = build_timetable(info)
+    candidates = build_candidates(info)
+    candidates.sort(key=lambda x: get_score(x, info), reverse=True)
+    candidates = candidates[:3]
+    for candidate in candidates:
+        time_table = TimeTable(title='table', year=info['year'], semester=info['semester'])
+        time_table.save()
+        time_table.lectures.set(candidate)
         recommends.append(time_table)
-    recommends.sort(key=lambda x: get_score(x, info, student), reverse=True)
-    print([get_score(x, info, student) for x in recommends])
-    for time_table in recommends:
-        time_table.delete()
-    return recommends[:num_recommends]
+
+    return recommends
+
+
+def build_candidates(info):
+    """
+    Build candidate lecture sets from seed sets.
+    Seed sets are courses/lectures that get high scores with regard to given user info.
+    :param info:
+    :return: candidates:
+    """
+    num_seeds = 10
+    seed_courses = []
+    courses = Course.objects.all()
+    for course in courses:
+        seed_courses.append(course)
+        if len(seed_courses) == num_seeds+1:
+            seed_courses.sort(key=lambda x: get_course_score(x, info), reverse=True)
+            seed_courses = seed_courses[:num_seeds]
+    print(seed_courses)
+
+    candidates = []
+    seed_lectures = Lecture.objects.filter(reduce(lambda x, y: x | y, [Q(course=c) for c in seed_courses]))
+    for lecture in seed_lectures:
+        candidate = branch_and_bound_help([lecture, ], lecture.course.credit, seed_lectures, info)
+        candidates.append(candidate)
+
+    return candidates
+
+
+def get_course_score(course, info):
+    """
+    Given course, calculates score for the course.
+    :param course:
+    :param info:
+    :return: score
+    """
+    score = 0
+    if course.department == info['student_department']:
+        score += 8
+    if course.grade == info['student_grade']:
+        score += 8
+    if course.type == '전필':
+        score += 4
+    if course.type == '전선':
+        score += 2
+    if course.type == '교양':
+        score += 1
+
+    return score
+
+
+max_expect = 0
+max_score = 0
+max_lectures = []
+
+
+def branch_and_bound_help(initial_lectures, initial_credit, seed_lectures, info):
+    """
+    This is a wrapper function for branch_and_bound.
+    :param initial_lectures:
+    :param initial_credit:
+    :param seed_lectures:
+    :param info:
+    :return max_lectures:
```

Comment:
Empty parameter, return value descriptions are meaningless.
It would be better to remove these lines and just leave function description.